micrometer-core/src/test/java/io/micrometer/core/instrument/binder/jvm/JvmGcMetricsTest.java

```diff
@@ -103,6 +116,57 @@ void gcTimingIsCorrect() {
         checkCollectionTime(pauseTimeMs, concurrentTimeMs);
     }
 
+    @Test
+    @Issue("gh-2872")
+    void sizeMetricsNotSetToZero() {
+        GcMetricsNotificationListener gcMetricsNotificationListener = binder.gcNotificationListener;
+        NotificationCapturingListener capturingListener = new NotificationCapturingListener();
+        Collection<Runnable> notificationListenerCleanUpRunnables = new ArrayList<>();
+
+        // register capturing notification listener
+        for (GarbageCollectorMXBean gcBean : ManagementFactory.getGarbageCollectorMXBeans()) {
+            if (!(gcBean instanceof NotificationEmitter)) {
+                continue;
+            }
+            NotificationEmitter notificationEmitter = (NotificationEmitter) gcBean;
+            notificationEmitter.addNotificationListener(capturingListener, notification -> notification.getType().equals(GarbageCollectionNotificationInfo.GARBAGE_COLLECTION_NOTIFICATION), null);
+            notificationListenerCleanUpRunnables.add(() -> {
+                try {
+                    notificationEmitter.removeNotificationListener(capturingListener);
+                } catch (ListenerNotFoundException ignore) {
+                }
+            });
+        }
+
+        try {
+            System.gc(); // capture real gc notifications
+            List<Notification> notifications = capturingListener.getNotifications();
+            assertThat(notifications).isNotEmpty();
```

Comment:
This is flaky because while the GC will be finished once `System.gc()` finishes, the corresponding notifications are asynchronous and may take arbitrary time to be processed by the configured `NotificationListener` instances. Unfortunately, using awaitility to wait until we have a notification is probably not the best approach since we know, for example, the issue fixed in this pull request happens with Shenandoah which sends 4 notifications (we want to capture all), but other collectors will have a different number of notifications per GC. While still potentially flaky and time consuming, perhaps a 100ms or so sleep is the best way to improve the flakiness here. Though again, if we only capture a portion of the notifications for the gc, the test isn't necessary checking that the issue is fixed.

Reply:
100 ms sleep added in 13fa32548e909f32e08f4b5f29d264093bd7612b